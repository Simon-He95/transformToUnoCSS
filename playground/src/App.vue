<script setup lang="ts">
import { computed, ref } from 'vue'
import { VividTyping } from 'vivid-typing'
import * as monaco from 'monaco-editor'
import HtmlWorker from 'monaco-editor/esm/vs/language/html/html.worker?worker'
import { copy, useAnimationFrame, useFocus } from 'lazy-js-utils'
import gitForkVue from '@simon_he/git-fork-vue'
import { useI18n } from 'vue-i18n'
import { transformVue } from '../../src/transformVue'
import { toUnocss } from '../../src/toUnocss'
import { isDark, toggleDark } from '~/composables'

const { t, locale } = useI18n()

const input = ref('')
let pre: any
  = '<template>\n  <button>button</button>\n</template>\n\n<style scoped>\n  button {\n    height: 32px;\n    display: flex;\n    justify-content: center;\n    align-items: center;\n    font-size: 14px;\n    cursor: pointer;\n    user-select: none;\n    padding: 8px 15px;\n    border-radius: 4px;\n    border: none;\n    box-sizing: border-box;\n    color: #fff;\n    background-color: #409eff;\n    margin: auto;\n  }\n  button:hover{\n    background-color: #67c23a ;\n  }\n</style>\n'

const transform = computed(() => toUnocss(input.value))
let editorComponent: any = null
const editor = ref(null)
const editorResult = ref<HTMLElement>()

const display = ref('')
const styleReg = /<style.*>(.*)<\/style>/s
const classReg = /(.*){/g

const editorInput = ref(`<template>
  <button>button</button>
</template>

<style scoped>
  button {
    height: 32px;
    display: flex;
    justify-content: center;
    align-items: center;
    font-size: 14px;
    cursor: pointer;
    user-select: none;
    padding: 8px 15px;
    border-radius: 4px;
    border: none;
    box-sizing: border-box;
    color: #fff;
    background-color: #409eff;
    margin: auto;
  }
  button:hover{
    background-color: #67c23a ;
  }
</style>
`)
monaco.editor.defineTheme('myTheme', {
  base: 'vs',
  inherit: true,
  rules: [{ background: 'EDF9FA', token: '' }],
  colors: {
    'editor.foreground': '#000000',
    'editor.background': '#EDF9FA',
    'editorCursor.foreground': '#8B0000',
    'editor.lineHighlightBackground': '#0000FF20',
    'editorLineNumber.foreground': '#008800',
    'editor.selectionBackground': '#88000030',
    'editor.inactiveSelectionBackground': '#88000015',
  },
})
monaco.editor.setTheme('myTheme')

onMounted(() => {
  useFocus('input') // 自动聚焦input

  self.MonacoEnvironment = {
    getWorker() {
      return new HtmlWorker()
    },
  }

  editorComponent = monaco.editor.create(editor.value!, {
    value: editorInput.value,
    fontFamily: 'Arial',
    fontSize: 20,
    language: 'html',
  })
  monaco.editor.create(editorResult.value!, {
    value: `<template>
  <button h-32px flex justify-center items-center text-14px cursor-pointer select-none px-15px py-8px border-rd-4px border-none box-border text-#fff bg-#409eff m-auto hover="bg-#67c23a">button</button>
</template>
<style scoped></style>
`,
    language: 'html',
    fontFamily: 'Arial',
    fontSize: 20,
    readOnly: true,
    acceptSuggestionOnEnter: 'smart',
  })
  display.value = codeToHtml(pre)
})

useAnimationFrame(async () => {
  const newInput = editorComponent!.getValue()
  if (!editorResult.value)
    return
  let code
  if ((!pre && newInput) || pre !== newInput) {
    pre = newInput

    try {
      code = await fetch('https://localhost/.netlify/functions/server', {
        method: 'POST',
        body: newInput,
      }).then(res => res.text())
    }
    catch (error) {
      code = await transformVue(newInput)
    }

    editorResult.value!.innerHTML = ''
    monaco.editor.create(editorResult.value!, {
      value: code,
      language: 'html',
      fontFamily: 'Arial',
      fontSize: 20,
      readOnly: true,
      acceptSuggestionOnEnter: 'smart',
    })
    display.value = codeToHtml(newInput)
  }
}, 200)

function codeToHtml(code: string) {
  return code
    .replace(styleReg, (all, v) =>
      all.replace(
        v,
        v.replace(
          classReg,
          (_: any, match: any) => `[data-v-display]${match} {`,
        ),
      ),
    )
    .replace('<template>', '')
    .replace('<\/template>', '')
}

const isCopy = ref(false)
const copyStyle = () => {
  if (copy(transform.value)) {
    isCopy.value = true
    window.parent.postMessage({ eventType: 'copy', text: transform.value })
  }

  setTimeout(() => {
    isCopy.value = false
  }, 1000)
}

const changelanguage = () => {
  if (locale.value === 'en')
    locale.value = 'zh'
  else locale.value = 'en'
}
</script>

<template>
  <div absolute flex="~ gap-2" z-2 left-2 top-2>
    <div
      hover="rotate-y-180deg "
      transition-all-800
      cursor-pointer
      @click="changelanguage"
    >
      <div i-fa:language />
    </div>
    <button
      class="icon-btn !outline-none border-none bg-transparent"
      @click="toggleDark()"
    >
      <div v-if="isDark" i-carbon-moon text-white />
      <div v-else i-carbon-sun />
    </button>
  </div>
  <gitForkVue
    link="https://github.com/Simon-He95"
    type="trapeziumType"
    position="right"
    content="Follow Me"
    color="pink"
  />
  <VividTyping
    content="Css To Unocss"
    animate-bounce-alt
    animate-delay-1500
    animate-count-infinite
    animate-duration-1s
    text-3xl
    color-pink
    py6
    text-center
    spilt-class="textshadow"
    class="typing"
    data-text="Css To Unocss"
  />
  <div h="100%" flex justify-center items-center flex-col p="y10">
    <input
      v-model="input"
      class="!outline-none"
      w="40%"
      text-4
      :placeholder="t('placeholder')"
      type="text"
      autocomplete="off"
      p="x6 y4"
      hover:border-pink
      border-1
    >
    <div v-if="transform" flex="~ gap-4" h-20 items-center>
      <h2>{{ t('result') }}</h2>
      <div flex gap-2 items-center>
        {{ transform }}
        <div
          :class="[
            isCopy
              ? 'i-carbon:checkmark-outline text-green!'
              : ' i-carbon:copy',
          ]"
          cursor-pointer
          hover="color-orange"
          @click="copyStyle"
        />
      </div>
    </div>
    <template v-else>
      <template v-if="input">
        <div h-20 v-html="t('issue')" />
      </template>
      <template v-else>
        <div h-20 box-border pt6 v-html="t('find')" />
      </template>
    </template>
  </div>
  <div flex>
    <div w="50%">
      <h1
        pl2
        class="textshadow"
        relative
        z-2
        :data-text="t('inputs')"
        indent-10
      >
        {{ t('inputs') }}
      </h1>
      <div ref="editor" h-100 />
    </div>
    <div w="50%">
      <h1
        pl2
        class="textshadow"
        relative
        z-2
        :data-text="t('outputs')"
        indent-10
      >
        {{ t('outputs') }}
      </h1>
      <div ref="editorResult" h-100 />
    </div>
  </div>
  <h1 pl2 class="textshadow" relative z-2 :data-text="t('render')" indent-10>
    {{ t('render') }}
  </h1>
  <div pb20 data-v-display v-html="display" />
</template>

<style scoped>
.textshadow::after {
  bottom: 0;
  color: #000;
  content: attr(data-text);
  -webkit-filter: blur(2px);
  filter: blur(2px);
  left: 0;
  -webkit-mask-image: linear-gradient(transparent, #000);
  mask-image: linear-gradient(transparent, #000);
  position: absolute;
  -webkit-transform: translate(-26px, 16px) skew(50deg) scaleY(0.6);
  transform: translate(-26px, 16px) skew(50deg) scaleY(0.6);
  z-index: 1;
}

.dark .textshadow::after {
  color: #fff;
}
</style>
