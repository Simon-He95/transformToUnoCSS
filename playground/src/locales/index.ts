import { createI18n } from 'vue-i18n'

const messages = {
  en: {
    placeholder: 'Enter the css value, such as: width:100rem...',
    result: 'Result:',
    match: 'The current value is wrong, or the matching error!',
    demo: 'If you enter a legal value, you can add your example here.',
    issue: `<h2>The current value is wrong, or the matching error</h2>
    If you enter a legal value, you can add your example here.
    <a
          href="https://github.com/Simon-He95/transformToUnocss/issues"
          target="_blank"
        >issues</a>`,
    find: `You can view the currently supported properties in <a
      href="https://github.com/Simon-He95/transformToUnocss"
      target="_blank"
    ><b>transform-to-unocss</b></a> and experience it.
<div>Already have vite plugin and cli version~</div>`,
    inputs: 'inputs:',
    outputs: 'outputs:',
    render: 'render:',
  },
  zh: {
    placeholder: '输入css值，如:width:100rem...',
    result: '结果:',
    issue: `<h2>当前值不对，或匹配错误！</h2>
    如果你输入的是一个合法的值，你可以再这里添加你的例子。
    <a
          href="https://github.com/Simon-He95/transformToUnocss/issues"
          target="_blank"
        >issues</a>`,
    find: ` 你可以在<a
      href="https://github.com/Simon-He95/transformToUnocss"
      target="_blank"
    ><b>transform-to-unocss</b></a>查看现在支持的属性，进行体验。
    <div>已有vite插件和cli版本~</div>`,
    inputs: '输入:',
    outputs: '输出:',
    render: '渲染:',
  },
}

const i18n = createI18n({
  legacy: false,
  globalInjection: true,
  locale: 'en',
  messages,
})

export default i18n
