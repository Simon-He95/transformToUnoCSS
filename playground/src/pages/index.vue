<script setup lang="ts"></script>

<template>
  <div style="background: red; width: 100%; line-height: 20px">
    <div class="red" />
  </div>
</template>

<style scoped>
div > .red {
  background-color: red;
  width: 100%;
}
.red {
  color: 'yellow';
}
.hi {
  aa: 'aa';
}
</style>
