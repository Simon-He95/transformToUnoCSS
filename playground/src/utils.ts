export const cssSuggestions = [
  // 文本属性
  'text-align',
  'text-align-last',
  'text-transform',
  'text-indent',
  'text-decoration',
  'text-shadow',
  'vertical-align',
  'white-space',
  'word-spacing',
  'word-break',
  'line-height',
  'letter-spacing',

  // 字体属性
  'font',
  'font-family',
  'font-size',
  'font-style',
  'font-variant',
  'font-weight',
  'font-stretch',

  // 颜色和背景
  'color',
  'opacity',
  'background',
  'background-color',
  'background-image',
  'background-repeat',
  'background-position',
  'background-size',
  'background-clip',
  'background-origin',
  'background-attachment',
  'background-blend-mode',

  // 边框和轮廓
  'border',
  'border-width',
  'border-style',
  'border-color',
  'border-radius',
  'border-top',
  'border-right',
  'border-bottom',
  'border-left',
  'outline',
  'outline-width',
  'outline-style',
  'outline-color',
  'outline-offset',

  // 尺寸和盒模型
  'width',
  'min-width',
  'max-width',
  'height',
  'min-height',
  'max-height',
  'box-sizing',
  'margin',
  'margin-top',
  'margin-right',
  'margin-bottom',
  'margin-left',
  'padding',
  'padding-top',
  'padding-right',
  'padding-bottom',
  'padding-left',

  // 显示和定位
  'display',
  'visibility',
  'position',
  'top',
  'right',
  'bottom',
  'left',
  'z-index',
  'float',
  'clear',

  // 柔性盒模型
  'flex',
  'flex-grow',
  'flex-shrink',
  'flex-basis',
  'flex-flow',
  'flex-direction',
  'flex-wrap',
  'justify-content',
  'align-items',
  'align-content',
  'order',
  'align-self',

  // 网格布局
  'grid',
  'grid-template-columns',
  'grid-template-rows',
  'grid-template-areas',
  'grid-auto-columns',
  'grid-auto-rows',
  'grid-auto-flow',
  'grid-column-gap',
  'grid-row-gap',
  'grid-gap',
  'grid-column-start',
  'grid-column-end',
  'grid-row-start',
  'grid-row-end',
  'grid-area',

  // 过渡和动画
  'transition',
  'transition-property',
  'transition-duration',
  'transition-timing-function',
  'transition-delay',
  'animation',
  'animation-name',
  'animation-duration',
  'animation-timing-function',
  'animation-delay',
  'animation-iteration-count',
  'animation-direction',
  'animation-fill-mode',
  'animation-play-state',

  // 变换
  'transform',
  'transform-origin',
  'transform-style',
  'perspective',
  'perspective-origin',
  'backface-visibility',

  // 滤镜
  'filter',

  // 其他
  'cursor',
  'list-style',
  'user-select',
  'resize',
  'overflow',
  'overflow-x',
  'overflow-y',
  'box-shadow',
  'isolation',
  'mix-blend-mode',
  'content',
  'counter-reset',
  'counter-increment',
]
