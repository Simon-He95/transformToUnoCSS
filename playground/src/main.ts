import routes from 'virtual:generated-pages'
import { createApp } from 'vue'
import { createRouter, createWebHistory } from 'vue-router'
import App from './App.vue'
import i18n from './locales'
// import '@unocss/reset/tailwind.css'
import './styles/main.css'
import 'uno.css'

const app = createApp(App)
const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
})
app.use(router)
app.use(i18n)
app.mount('#app')
