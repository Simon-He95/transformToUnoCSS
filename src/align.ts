import { getLastName, transformImportant } from './utils'

export function align(key: string, val: string) {
  const [value, important] = transformImportant(val)
  return `${getLastName(key)}-${getLastName(value)}${important}`
}
