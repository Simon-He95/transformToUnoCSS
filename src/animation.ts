import { getVal, transformImportant } from './utils'

export function animation(key: string, val: string) {
  const [value, important] = transformImportant(val)

  if (key === 'animation-delay')
    return `animate${getVal(value)}${important}`
  if (key === 'animation')
    return `animate-${value.split(' ')[0]}${important}`
  return `animate-${value}${important}`
}
