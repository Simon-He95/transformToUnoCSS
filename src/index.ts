import { getFirstName } from './utils'
import { transformSize } from './size'
import { transformBackground } from './background'
import { transformMax } from './max'
import { transformFont } from './font'
import { transformCursor } from './cursor'
import { transformMargin } from './margin'
import { transformOpacity } from './opacity'
import { transformPosition } from './position'

const typeMap: any = {
  background: transformBackground,
  width: transformSize,
  height: transformSize,
  font: transformFont,
  max: transformMax,
  cursor: transformCursor,
  margin: transformMargin,
  padding: transformMargin,
  opacity: transformOpacity,
  position: transformPosition,
}
export function transformToUnocss(css: String) {
  const splitReg = /([\w-]+)\s*:\s*([.\w\(\)-\s%+'",]+)/
  const match = css.match(splitReg)
  if (!match)
    return
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  const [_, key, val] = match
  const first = getFirstName(key)
  const result = typeMap[first](key, val)
  return result
}

