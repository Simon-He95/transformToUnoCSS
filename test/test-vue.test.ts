import fsp from 'fs/promises'
import path from 'path'
import { describe, expect, it } from 'vitest'
import { transfromCode } from '../src'



describe('test-vue', async () => {
  const demo = await fsp.readFile('./test/demo/test-vue.vue', 'utf-8')
  const filepath = path.resolve(process.cwd(), './test/demo/test-vue.vue')
  it('test-vue.vue', async () => {
    expect(await transfromCode(demo, {filepath, type:'vue'})).toMatchSnapshot()
  })
})
