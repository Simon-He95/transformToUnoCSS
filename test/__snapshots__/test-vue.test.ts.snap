// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test-vue > test-vue.vue 1`] = `
"<script setup lang="ts">
// 这里可以添加你的逻辑
</script>
<template>
  <div class="portal-container min-h-100vh flex flex-col font-[Arial,_sans-serif] bg-[#f5f6fa]">
    <header class="portal-header bg-[#2d8cf0] text-[#fff] px-32px py-24px flex flex-col items-start">
      <h1>门户首页</h1>
      <nav class="mt-12px">
        <a class="text-[#fff] mr-20px decoration-none font-bold hover-decoration-underline" href="#">首页</a>
        <a class="text-[#fff] mr-20px decoration-none font-bold hover-decoration-underline" href="#">功能一</a>
        <a class="text-[#fff] mr-20px decoration-none font-bold hover-decoration-underline" href="#">功能二</a>
        <a class="text-[#fff] mr-20px decoration-none font-bold hover-decoration-underline" href="#">关于</a>
      </nav>
    </header>
    <main class="portal-main flex-1 px-32px py-40px">
      <section class="welcome-section mb-32px">
        <h2>欢迎来到门户页面</h2>
        <p>这里是你的 Vue 门户页面示例。</p>
      </section>
      <section class="features-section">
        <h3>功能介绍</h3>
        <ul class="pl-20px">
          <li>功能一描述</li>
          <li>功能二描述</li>
          <li>更多功能...</li>
        </ul>
      </section>
    </main>
    <footer class="portal-footer bg-[#222] text-[#fff] text-center px-0 py-16px text-14px">
      &copy; 2025 门户页面示例
    </footer>
  </div>
</template>
<style scoped></style>
"
`;
